Adds retry mechanism for Node resource

msherif1234 opened the pull request:
Leverages the same retry mechanism used by #2809 to build unified retry logic lib and use it for node and network policy resources
Signed-off-by: Mohamed Mahmoud <mmahmoud@redhat.com>

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1253,11 +1254,12 @@ func (oc *Controller) deleteNodeLogicalNetwork(nodeName string) error {
 	// Clean up as much as we can but don't hard error
 	for _, hostSubnet := range hostSubnets {
 		if err := oc.deleteNodeHostSubnet(nodeName, hostSubnet); err != nil {
 			klog.Errorf("Error deleting node %s HostSubnet %v: %v", nodeName, hostSubnet, err)
```
we dont need to klog these now that you changed it to return an error. We can just return the error and log it from the caller.

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1481,7 +1488,9 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
 
-		oc.deleteNode(nodeSwitch.Name, subnets)
+		if err := oc.deleteNode(nodeSwitch.Name, subnets); err != nil {
+			klog.Warningf("Failed to delete node:%s, err:%v", nodeSwitch.Name, err)
```
if we fail to delete a node shouldn't we return an error here?

msherif1234 replied:
this one was in syncNodes() path so I wasn't sure if I wanted to return error here

trozet replied:
yeah we will retry sync nodes if it fails. I think this should be an error. wdyt @flavio-fernandes ?

flavio-fernandes replied:
I agree. We should ensure all these calls are idem-potent, tho:
```
func (oc *Controller) deleteNode(nodeName string, hostSubnets []*net.IPNet) {
	// Clean up as much as we can but don't hard error
	for _, hostSubnet := range hostSubnets {
		if err := oc.deleteNodeHostSubnet(nodeName, hostSubnet); err != nil {
			klog.Errorf("Error deleting node %s HostSubnet %v: %v", nodeName, hostSubnet, err)
		} else {
			util.UpdateUsedHostSubnetsCount(hostSubnet, &oc.v4HostSubnetsUsed, &oc.v6HostSubnetsUsed, false)
		}
	}
	// update metrics
	metrics.RecordSubnetUsage(oc.v4HostSubnetsUsed, oc.v6HostSubnetsUsed)

	if err := oc.deleteNodeLogicalNetwork(nodeName); err != nil {
		klog.Errorf("Error deleting node %s logical network: %v", nodeName, err)
	}

	if err := oc.gatewayCleanup(nodeName); err != nil {
		klog.Errorf("Failed to clean up node %s gateway: (%v)", nodeName, err)
	}

	if err := oc.joinSwIPManager.ReleaseJoinLRPIPs(nodeName); err != nil {
		klog.Errorf("Failed to clean up GR LRP IPs for node %s: %v", nodeName, err)
	}

	if err := libovsdbops.DeleteNodeChassis(oc.sbClient, nodeName); err != nil {
		klog.Errorf("Failed to remove the chassis associated with node %s in the OVN SB Chassis table: %v", nodeName, err)
	}
}
```

trozet replied:
@msherif1234 I think you need to turn the warning into the error message, and like Flavio said, ensure the things called in delete node are idempotent

msherif1234 replied:
done

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1500,157 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+		return err
+	}
+
+	klog.V(5).Infof("Added event for Node %q", node.Name)
```
"Adding node %q"

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1047,148 +1064,54 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
-			}
-			if !shouldUpdate {
-				// the hostsubnet is not assigned by ovn-kubernetes
+			if err := oc.updateNodeObj(node, oldNode); err != nil {
```
need a checkandSkip before this

trozet replied:
before executing updateNodeObj we might need to fetch the newest version of node. In pod retry we do this, in policy we don't (might need to change that). The reason is in retry goroutine we fetch the latest version of the object. That means we could have revisions A,B,C (C is current). A fails to get added, goes into retry. Retry loop executes and adds C. Meanwhile B was added to the queue and then gets processed here. We dont want to update to B here. We want to get C and realize there is nothing to change.

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1047,148 +1064,54 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
-				// the hostsubnet is not assigned by ovn-kubernetes
+			if err := oc.updateNodeObj(node, oldNode); err != nil {
+				klog.Errorf("Failed to update node %s, error: %v", node.Name, err)
+				oc.unSkipRetryNode(node)
```
need an initRetryNode before this

trozet commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,129 @@
+
+// initRetryObj tracks a failed object to potentially retry later
+// initially it is marked as skipped for retry loop (ignore = true)
+func (oc *Controller) initRetryObj(obj interface{}, lock *sync.Mutex, retryObjs interface{}, key string) {
```
retryObjs can still be a map[string]interface right? You don't need to make the entire map an interface?

trozet commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,140 @@
+	defer (*lock).Unlock()
+	v := reflect.ValueOf(retryObjs)
+	if v.MapIndex(reflect.ValueOf(key)).IsValid() {
+		entry = v.MapIndex(reflect.ValueOf(key)).Interface().(*retryObjEntry)
```
if the value that the map holds is always a retryObjEntry, it doesn't need to be an interface. Only the retryObjEntry->newObj (etc) needs to be an interface because it can hold different kapi types. Change the map to map[string]*retryObjEntry

msherif1234 commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,140 @@
+
+// initRetryObj tracks a failed object to potentially retry later
+// initially it is marked as skipped for retry loop (ignore = true)
+func (oc *Controller) initRetryObj(obj interface{}, lock *sync.Mutex, retryObjs map[string]interface{}, key string) {
```
great idea !!

msherif1234 commented on go-controller/pkg/ovn/policy_retry.go:
```diff
@@ -1,199 +0,0 @@
-package ovn
```
merged all commit in one since there are lots of overlapping files

ricky-rav commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1047,148 +1072,80 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
-				// delete stale chassis in SBDB if any
-				oc.deleteStaleNodeChassis(node)
+				// deletion was a success; remove delete retry entry
+				oc.retryNodes.removeDeleteFromRetryObj(newNode.Name)
```
should we call it on `newNode.Name` or `oldNode.Name`? Here we've just successfully deleted `oldNode`.

msherif1234 replied:
the remove is for newNode but all the logic above to see if we have stale obj should have been done for newNode not oldNode I renamed the var last night to make it clear but I incorrectly used them I will fix it

ricky-rav commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1499,233 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+		mgmtPortFailed.Store(node.Name, true)
+		gatewaysFailed.Store(node.Name, true)
+		errs = append(errs, err)
+	}
```
if `oc.addNode(node)` fails, should we really go on with the following calls (`oc.syncNodeClusterRouterPort`, `oc.syncNodeManagementPort`, etc.)? I see that the original code returned if `addNode` failed.

msherif1234 replied:
u are right I should have left those early 2 return and just aggregate the other cases

ricky-rav commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1499,233 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+	_, failed := addNodeFailed.Load(node.Name)
+	if failed {
+		hostSubnets, err = oc.addNode(node)
+		if err != nil {
```
should we add the node back to `addNodeFailed` in case it fails again to be added?

msherif1234 replied:
the original code didn't added it and return so I kept the same behavior
```golang
if failed {
		hostSubnets, err = oc.addNode(node)
		if err != nil {
			klog.Errorf("NodeUpdate: error creating subnet for node %s: %v", node.Name, err)
			return
		}
		addNodeFailed.Delete(node.Name)
```

ricky-rav commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,118 @@
+
+	"k8s.io/klog/v2"
+)
+
```
Would it make sense to have the definition of `retryObjs` and `retryObjEntry` here?

ricky-rav commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,118 @@
+}
+
+// initRetryWithDelete tracks a failed object to potentially retry later
+// initially it is marked as skipped for retry loop (ignore = true)
```
This is the same comment as in `initRetryObj` where we set `newObj`, while here we set `oldObj`. This is more of placeholder for me to think of a better comment for this function :)

msherif1234 replied:
I fixed the comment to make it more clear

ricky-rav commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1499,233 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+	var gatewaysFailed sync.Map
+	var mgmtPortFailed sync.Map
+	var addNodeFailed sync.Map
+	var nodeClusterRouterPortFailed sync.Map
```
In the original code, I see that these sync.Maps are defined once in:
```
func (oc *Controller) WatchNodes() {
	var gatewaysFailed sync.Map
	var mgmtPortFailed sync.Map
	var addNodeFailed sync.Map
	var nodeClusterRouterPortFailed sync.Map
```
... and then they are accessed, possibly concurrently, by the ResourceEventHandlerFuncs defined soon afterwards. 

Here we're redefining these sync.Maps locally in each `addNodObj`, `updateNodeObj` and deleteNodeObj`. Is the result equivalent?I have zero experience with `sync.Map`, so this is more a question for my knowledge than for this PR... :)

msherif1234 replied:
I think you are right its up to k8s how it invoke those handlers and they could be concurrent that is probably the reason for those syncMap orginally I will fix my changes

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1497,226 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+			// create new node if needed
+			if node != nil {
+				klog.Infof("Node Retry: Creating new node for %s", nodeName)
+				if err := oc.addNodeObj(node); err != nil {
```
right it will Store new failures, but it wont remove failures if it succeeds. So you could have:
1. Add node fail: one or more of these maps is updated with a failure = true. InitRetry
2. Sometime later, retry happens: calls addNodeObj. Now how do we clear out the failures from the sync maps?

msherif1234 replied:
during iterateRetryNodes before recreate again I can clean old sync maps

ricky-rav replied:
As I see it, `iterateRetryNodes` [removes a node from the maps ](https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-1f90f4648dc7d68c42854549c4257ddcfc9bd9a5b6fc480c74e3a23989914bc3R1703-R1706) and `addNodeObj`, if it fails, [adds it to the maps](https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-1f90f4648dc7d68c42854549c4257ddcfc9bd9a5b6fc480c74e3a23989914bc3R1516-R1520).

trozet replied:
I dont think this will work. The retry action is always an "addNode", which doesn't take into consideration the previous failure could have been an update. I think the easiest way is to consolidate the add/update functions into one. If you want I can push a commit showing how it would look.

flavio-fernandes commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1492,3 +1495,80 @@ func (oc *Controller) shutdownHandlers(np *networkPolicy) {
+		}
+		backoff := (npEntry.backoffSec * time.Second) + (time.Duration(rand.Intn(500)) * time.Millisecond)
+		npTimer := npEntry.timeStamp.Add(backoff)
+		if updateAll || now.After(npTimer) {
```
NIT: It may save us the long indented block if we rewrote this block as:
```
if !updateAll && now.Before(npTimer) {
    klog.V(5).Infof("%s retry network policy not after timer yet, time: %s", namespacedName, npTimer)
    continue
}

... <the original block, now not needed to be inside long "if { }" block
```

msherif1234 replied:
original code not me :)

flavio-fernandes commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1492,3 +1495,80 @@ func (oc *Controller) shutdownHandlers(np *networkPolicy) {
+				npEntry.oldObj = nil
+			}
+
+			// create new policy if needed
```
`// check if we need to create anything`  :)

flavio-fernandes commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1492,3 +1495,80 @@ func (oc *Controller) shutdownHandlers(np *networkPolicy) {
+				klog.Infof("%s policy not found in the informers cache, not going to retry policy create", namespacedName)
+				npEntry.newObj = nil
+			} else {
+				policyToCreate = np
```
NIT: Is there any value in updating `npEntry.newObj = np` so we have the latest and greatest copy of it?

msherif1234 replied:
not sure about this one I will defer to @trozet  since this is his code I just moved to policy.go as is

flavio-fernandes commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,128 @@
+
+// initRetryWithDelete tracks an object that failed to be deleted to potentially retry later
+// initially it is marked as skipped for retry loop (ignore = true)
+func (r *retryObjs) initRetryObjWithDelete(obj interface{}, key string, config interface{}) {
```
Can we combine these 2 functions initRetryObjWithDelete + initRetryObj ? 
Or maybe rename initRetryObj to initRetryObjWithAdd ?
Feels like they do almost exaclt the same think besides the `oldObj` vs `newObj` part.

msherif1234 replied:
that is exactly right one handle add path and another for delete I can rename with Add prefix

flavio-fernandes commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,128 @@
+)
+
+type retryObjEntry struct {
+	newObj     interface{}
```
is there ever a case when both newObj and oldObj are set?
If not, could this just be a obj and a `bool isDelete` ?
Please forgive and ignore me if thatis a silly idea. ;)

msherif1234 replied:
during update ? u can have old and new objects that need to be taken care of see the logic in the watch code path in ovn.go

flavio-fernandes replied:
ah. ack. thanks for the clarification.

flavio-fernandes reviewed (commented):
@msherif1234 I added a few comments in this PR. Please keep in mind they are mostly question and suggestions. So feel free to ignore them. ;)

msherif1234 commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,128 @@
+
+// getObjRetryEntry returns a copy of an object  retry entry from the cache
+// Used for Unit Tests only
+func (r *retryObjs) getObjRetryEntry(key string) *retryObjEntry {
```
what about using testprefix in the name ?

msherif1234 commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,128 @@
+// addDeleteToRetryObj adds an old object that needs to be cleaned up to a retry object
+// includes the config object as well in case the namespace is removed and the object is orphaned from
+// the namespace
+func (r *retryObjs) addDeleteToRetryObj(obj interface{}, key string, config interface{}) {
```
good question, the functions with return their return of no use atm, so I am removing the return code since its not been used

msherif1234 commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1096,7 +1098,8 @@ func (oc *Controller) addNetworkPolicy(policy *knet.NetworkPolicy) error {
 			// rollback failed, add to retry to cleanup
-			oc.addDeleteToRetryPolicy(policy, np)
+			key := getPolicyNamespacedName(policy)
+			oc.retryNetPolices.addDeleteToRetryObj(policy, key, np)
```
if u check the code path where this function is called there is always a call to `oc.retryNetPolices.initRetryObj` to make sure we have an entry in retry map

ricky-rav commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1047,148 +1041,66 @@ func (oc *Controller) syncNodeGateway(node *kapi.Node, hostSubnets []*net.IPNet)
-					klog.Errorf("NodeUpdate: error creating subnet for node %s: %v", node.Name, err)
+			newNode := new.(*kapi.Node)
+			oc.retryNodes.checkAndSkipRetryObj(newNode.Name)
+			if retryEntry := oc.retryNodes.getObjRetryEntry(newNode.Name); retryEntry != nil && retryEntry.oldObj != nil {
```
In the existing code for network policies, in this part we actually consider and process the old object:
https://github.com/ovn-org/ovn-kubernetes/pull/2870/files#diff-796adb9d8a9b5b5636dc81e7350a00ef143e6e5bcb52ec30dc350c866cd9208cR677-R686

Here I see that you are doing everything around the new node. Either one of the two implementations is wrong, I suppose :)

msherif1234 replied:
I was more of less based off the pods handling since it does have update path unlike netpol

msherif1234 replied:
nodes is more like pods with update hence I was more aligned with pod update https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L588

trozet replied:
I think Ricky means that a few lines down:
```
if err := oc.deleteNodeEvent(newNode); err != nil {
```

^ that should be deleting the old node object, not the new

msherif1234 replied:
understood but this isn't what pod update do 
https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/ovn.go#L607
its a node that was in cache for delete retry hence we redelete it

trozet reviewed (changes requested):
I think this is getting too complicated and I struggled to write a github comment that would clarify what I'm looking for. I thought it was just going to add more confusion, so I just added a commit on top of yours that I think simplifies the code here:
https://github.com/trozet/ovn-kubernetes/commit/942c3b3fc5582549328cb2bbb2e8c270eb424f17

msherif1234 commented:
> I think this is getting too complicated and I struggled to write a github comment that would clarify what I'm looking for. I thought it was just going to add more confusion, so I just added a commit on top of yours that I think simplifies the code here: [trozet@942c3b3](https://github.com/trozet/ovn-kubernetes/commit/942c3b3fc5582549328cb2bbb2e8c270eb424f17)

at first glance I didn't like the additional parameters but I realized checking the syncMap was short for at least one case, then I immediately realized that ur scheme give much better control over what need to get updated so I switched to ur changes my only concer the change u did to `clearInitialNodeNetworkUnavailableCondition` where u dropped the following logic
```golang
// if newNode is not nil, then we are called from UpdateFunc()
	if newNode != nil && reflect.DeepEqual(origNode.Status.Conditions, newNode.Status.Conditions) {
		return
	}
```

trozet commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,178 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+			oc.mgmtPortFailed.Store(node.Name, true)
+			oc.gatewaysFailed.Store(node.Name, true)
+			return fmt.Errorf("nodeAdd: error creating subnet for node %s: %w", node.Name, err)
+		}
```
need oc.addNodeFailed.Delete(node.Name)

trozet commented on go-controller/pkg/ovn/master_test.go:
```diff
@@ -994,7 +994,16 @@ var _ = ginkgo.Describe("Gateway Init Operations", func() {
 			err = clusterController.syncGatewayLogicalNetwork(updatedNode, l3GatewayConfig, []*net.IPNet{subnet}, hostAddrs)
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
-
+			clusterController.retryNodes.initRetryObjWithAdd(testNode, testNode.Name)
```
could we add a more robust test case that actually forces a node to get retried? There are a couple examples with network policy.

msherif1234 replied:
I looked at that initally node object unlike others where I can dyn add/delete calling the client apis I will look at this more in the following PRs and see what can possibly done for this resource

trozet commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,138 @@
+type retryObjEntry struct {
+	newObj     interface{}
+	oldObj     interface{}
+	config     interface{}
```
so this is a generic config object that resource specfic handlers may use to store internal information. Maybe we should add a comment here indicating that.

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -633,95 +630,92 @@ func (oc *Controller) WatchPods() {
-						getPolicyNamespacedName(policy), err)
-					return
+				knp, ok := retryEntry.oldObj.(*kapisnetworking.NetworkPolicy)
+				if ok {
```
this should never fail and if somehow it does here, we are just ignoring it. I think you should just change this to a direct assertion and if it fails we would panic. There's something very wrong with our code if we are storing wrong types. So just do knp := retryEntry.oldObj.(*kapisnetworking.NetworkPolicy)

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -633,95 +630,92 @@ func (oc *Controller) WatchPods() {
-						return
+				if retryEntry := oc.retryNetPolices.getObjRetryEntry(oldKey); retryEntry != nil && retryEntry.oldObj != nil {
+					knp, ok := retryEntry.oldObj.(*kapisnetworking.NetworkPolicy)
+					if ok {
```
same here, and anywhere else this pattern exists

trozet commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1492,3 +1495,80 @@ func (oc *Controller) shutdownHandlers(np *networkPolicy) {
+			var np *networkPolicy
+			klog.Infof("Network Policy Retry: Removing old policy for %s", namespacedName)
+			knp := npEntry.oldObj.(*knet.NetworkPolicy)
+			if _, ok := npEntry.config.(*networkPolicy); !ok {
```
with this logic, np will always be nil. Need to fix this.

trozet reviewed (changes requested):
@tssurya PTAL

trozet reviewed (approved):
/lgtm

@tssurya PTAL

tssurya commented on go-controller/pkg/ovn/policy.go:
```diff
@@ -1492,3 +1495,80 @@ func (oc *Controller) shutdownHandlers(np *networkPolicy) {
+			continue
+		}
+		// check if we need to create
+		var policyToCreate *knet.NetworkPolicy
```
same as I mentioned for nodesretry, do we need a new variable? can't we reuse the newObj itself?

tssurya commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// checkAndSkipRetryObj sets a specific entry from the map to be ignored for subsequent retries
+func (r *retryObjs) checkAndSkipRetryObj(key string) {
```
do we really need "check"? feels overloaded, since we have `unSkipRetryObj` itsn't `skipRetryObj` enough? its implicit that if the value doesn't exist in the map nothing will be done anyways,

msherif1234 replied:
I am keeping backward comptability with existing implementation

tssurya replied:
why is that? isn't this the fresh start of a standard set of retryObjs? Isn't it better to be precise with function namings? there is no backwards compatibility with namings...not sure what you mean by "backward compatibility" here...

tssurya commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,139 @@
+	if entry, ok := r.entries[key]; ok {
+		entry.timeStamp = time.Now()
+		entry.oldObj = obj
+		if entry.config == nil {
```
haven't had a chance to check the namespace.go and policy.go in depth, but we are 100% sure we don't want to overwrite an existing config unlike oldObj right? like I understand why this check is here because we call config=nil from most of the places, just wondering if 
```
oc.retryNetPolices.initRetryObjWithDelete(np.policy, key, np)
```
is a one time thing? I guess answer is yes.

msherif1234 replied:
u mean one time for every delete event right ? if so then yes

tssurya commented on go-controller/pkg/ovn/master_test.go:
```diff
@@ -1214,6 +1224,389 @@ var _ = ginkgo.Describe("Gateway Init Operations", func() {
+			clusterController.WatchNodes()
+			ginkgo.By("injecting node add errors")
+			// inject errors
+			clusterController.addNodeFailed.Store(testNode.Name, true)
```
do we need to do this on L1377 and L1378? shouldn't we be actually verifying if addUpdateNodeEvent() sets these correctly instead of us adding it to the failed maps since the db was simulated to act like its not reachable?

msherif1234 replied:
the test hit the update path not the create so if I don't inject error we won't see any failure in addUpdate which leads to no retries

tssurya reviewed (commented):
I don't see any major glitches in logic. lgtm.
questions/suggestions/comments inline.
Thanks!

ricky-rav commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1481,7 +1481,9 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
 		}
 
-		oc.deleteNode(nodeSwitch.Name, subnets)
+		if err := oc.deleteNode(nodeSwitch.Name, subnets); err != nil {
```
`syncNodesRetriable` makes the sync function of the node watcher execute again if it fails. Here we're making the addFunc, deleteFunc and updateFunc execute again if they fail. So yeah, it's a different beast!

msherif1234 replied:
correct

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+				klog.Infof("%s node not found in the informers cache, not going to retry node setup", nodeName)
+				entry.newObj = nil
+			} else {
+				node = kNode
```
no we need to set node so it can be used down if creation failed and we need to recreate

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+			}
+
+			// create new node if needed
+			if node != nil {
```
no because we didn't update newobj to latest but after change based on prev comment we can use newObj here

msherif1234 commented on go-controller/pkg/ovn/obj_retry.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+// checkAndDeleteRetryObj deletes a specific entry from the map
+func (r *retryObjs) checkAndDeleteRetryObj(key string, withLock bool) {
```
checking if the key exists which is embedded in `delete()` also keeping the same naming convention as it was before

tssurya replied:
I think we should just call this deleteRetryObj... no need to hint the underlying map delete functionality in the call. In pods retry where this originally came from, we used to actually check if it exists... we don't need that anymore.

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+	klog.V(5).Infof("Deleting Node %q. Removing the node from "+
+		"various caches", node.Name)
+
+	nodeSubnets, _ := util.ParseNodeHostSubnetAnnotation(node)
```
doesn't that mean we have bigger issue if we are at delete node and it has not subnet annotation ? that is probably why the current code didn't check and proceeded to node delete ? I can add it but that is problem for node create note delete and retry delete won't help WDYT ?

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+			oc.gatewaysFailed.Store(node.Name, true)
+			return fmt.Errorf("nodeAdd: error creating subnet for node %s: %w", node.Name, err)
+		}
+		oc.addNodeFailed.Delete(node.Name)
```
correct this function as the name indicate handle both add and update

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+	syncClusterRouter bool
+	syncMgmtPort      bool
+	syncGw            bool
+}
```
well those flags also can be all set to true on fresh create during update failure or iterate they are set based on syncMap as u indicated to control which part need to be retried

msherif1234 commented on go-controller/pkg/ovn/master.go:
```diff
@@ -1491,3 +1493,177 @@ func (oc *Controller) syncNodesRetriable(nodes []interface{}) error {
+	}
+
+	// delete stale chassis in SBDB if any
+	oc.deleteStaleNodeChassis(node)
```
what do u mean the retry code will retry this as its not controller by a flag ?

trozet replied:
I think she means it doesn't return an error type, so how do we know if it was successful or not. It's a good point, right now that function just logs an error, perhaps it should return an error and we propagate that up so it will get retried.

msherif1234 commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1121,74 +1078,48 @@ func (oc *Controller) WatchNodes() {
+			gwSync := failed || gatewayChanged(oldNode, newNode) || nodeSubnetChanged(oldNode, newNode) || hostAddressesChanged(oldNode, newNode)
+
+			if err := oc.addUpdateNodeEvent(newNode,
+				&nodeSyncs{nodeSync, clusterRtrSync, mgmtSync, gwSync}); err != nil {
```
correct we only set true in syncMap and we check if there if it exists to drive the failed flag

ricky-rav commented:
/lgtm

trozet commented on go-controller/pkg/ovn/ovn.go:
```diff
@@ -1121,74 +1078,48 @@ func (oc *Controller) WatchNodes() {
+				oc.retryNodes.unSkipRetryObj(newNode.Name)
+				return
 			}
+			oc.retryNodes.checkAndDeleteRetryObj(newNode.Name, true)
```
the processing of events is serialized by its key into only one thread at a time. So even in the homegrown federated queue only a single namespace_name pod or node event type could be processed by a worker at a time. Same thing for workqueues.

tssurya reviewed (approved):
/lgtm
I still wish for us to call the functions `skipRetryObj` and `unSkipRetryObj` and `deleteRetryObj` without adding "check" to everything since the logic has changed, but I won't stall this just for that
@ricky-rav : you can perhaps do that when you submit your PR.

thanks @msherif1234 !

msherif1234 commented:
/retest-failed
